spring-integration-core/src/test/java/org/springframework/integration/handler/AsyncHandlerTests.java

```diff
@@ -0,0 +1,257 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.handler;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.integration.channel.DirectChannel;
+import org.springframework.integration.channel.QueueChannel;
+import org.springframework.integration.endpoint.EventDrivenConsumer;
+import org.springframework.integration.gateway.GatewayProxyFactoryBean;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public class AsyncHandlerTests {
+
+	private final QueueChannel output = new QueueChannel();
+
+	private AbstractReplyProducingMessageHandler handler;
+
+	private volatile CountDownLatch latch;
+
+	private volatile int whichTest;
+
+	private volatile Exception failedCallbackException;
+
+	private volatile CountDownLatch exceptionLatch = new CountDownLatch(1);
+
+	private volatile CountDownLatch noOutputChannelLatch = new CountDownLatch(1);
+
+	@Before
+	public void setup() {
+		this.handler = new AbstractReplyProducingMessageHandler() {
+
+			@Override
+			protected Object handleRequestMessage(Message<?> requestMessage) {
+				final SettableListenableFuture<String> future = new SettableListenableFuture<String>();
+				Executors.newSingleThreadExecutor().execute(new Runnable() {
+
+					@Override
+					public void run() {
+						try {
+							latch.await(10, TimeUnit.SECONDS);
+							switch (whichTest) {
+								case 0:
+									try {
+										future.set("reply");
+									}
+									catch (Exception e) {
```

Comment:
This doesn't work with SF-4.3 - `ListenableFutureCallbackRegistry`:
```
private void notifySuccess(SuccessCallback<? super T> callback) {
	try {
		callback.onSuccess((T) this.result);
	}
	catch (Throwable ex) {
		// Ignore
	}
}
```
Investigating...